Adjoint metric tensor (#1992)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

<h3>New features since last release</h3>

* Added the adjoint method for the metric tensor.
  [(#1992)](https://github.com/PennyLaneAI/pennylane/pull/1992)

  This method, detailed in [Jones 2020](https://arxiv.org/abs/2011.02991),
  computes the metric tensor using four copies of the state vector and
  a number of operations that scales quadratically in the number of trainable

>   parameters. As it makes use of state cloning it is inherently classical

Reviewer: ```suggestion
  parameters. As it makes use of state cloning, it is inherently classical
```

  and to be used on state vector simulators only.

>   It is particular useful for larger circuits for which back-propagation requires

Reviewer: ```suggestion
  It is particular useful for larger circuits for which backpropagation requires
```

>   inconvenient or even unfeasible amounts of storage, but is slower.

Reviewer: but is slower than backpropagation?

Does that mean your backprop proof of concept function in the unittests is faster?

Author: It seems that way :)

Using this version:
```python
def autodiff_metric_tensor(ansatz, num_wires):
    """Compute the metric tensor by full state vector
    differentiation via autograd."""
    dev = qml.device("default.qubit", wires=num_wires)

    @qml.qnode(dev)
    def qnode(*params):
        ansatz(*params, wires=dev.wires)
        return qml.state()
    
    def mt(*params):
        state = qnode(*params)
        rqnode = lambda *params: pnp.real(qnode(*params))
        iqnode = lambda *params: pnp.imag(qnode(*params))
        rjac = qml.jacobian(rqnode)(*params)
        ijac = qml.jacobian(iqnode)(*params)
        if isinstance(rjac, tuple):
            out = []
            for rc, ic in zip(rjac, ijac):
                c = rc + 1j * ic
                psidpsi = pnp.tensordot(pnp.conj(state), c, axes=([0], [0]))
                out.append(
                    pnp.real(
                        pnp.tensordot(pnp.conj(c), c, axes=([0], [0]))
                        - pnp.tensordot(pnp.conj(psidpsi), psidpsi, axes=0)
                    )
               )
            return tuple(out)
        jac = rjac + 1j * ijac
        psidpsi = pnp.tensordot(pnp.conj(state), jac, axes=([0], [0]))
        return pnp.real(
            pnp.tensordot(pnp.conj(jac), jac, axes=([0], [0]))
            - pnp.tensordot(pnp.conj(psidpsi), psidpsi, axes=0)
        )

    return mt
```
and this circuit:
```python

dev = qml.device("default.qubit.autograd", wires=4)

def ansatz(x, y, wires):
    qml.Rot(*x[0], wires=0)
    qml.Rot(*x[1], wires=1)
    qml.Rot(*x[2], wires=2)
    qml.CNOT(wires=[0, 1])
    qml.CNOT(wires=[1, 2])
    qml.CNOT(wires=[2, 0])
    qml.RY(y[0], wires=0)
    qml.RY(y[1], wires=1)
    qml.RY(y[0], wires=2)

@qml.qnode(dev, interface="autograd")
def circuit(x, y):
    ansatz(x, y, dev.wires)
    return qml.expval(qml.PauliZ(0)@qml.PauliZ(2))

x = pnp.array([[0.2, 0.4, -0.1], [-2.1, 0.5, -0.2], [0.1, 0.7, -0.6]])
y = pnp.array([1.3, 0.2], requires_grad=True)

adj_mt_fn = qml.adjoint_metric_tensor(circuit)
mt_fn = qml.metric_tensor(circuit)
print("Adjoint:")
%timeit adj_mt_fn(x, y)
print("Cov_mat+Hadamards:")
%timeit mt_fn(x, y)
print("Backprop:")
%timeit autodiff_metric_tensor(ansatz, 3)(x, y)
```
I get the times
```pycon
Adjoint:
116 ms ± 11.6 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
Cov_mat+Hadamards:
174 ms ± 9.43 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
Backprop:
54.3 ms ± 1.82 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reviewer: huh 🤔 I suppose that makes sense

  Furthermore, the adjoint method is only available for analytic computation, not
  for measurements simulation with `shots!=None`.

  ```python
  dev = qml.device("default.qubit", wires=3)
  
  @qml.qnode(dev)
  def circuit(x, y):
      qml.Rot(*x[0], wires=0)
      qml.Rot(*x[1], wires=1)
      qml.Rot(*x[2], wires=2)
      qml.CNOT(wires=[0, 1])
      qml.CNOT(wires=[1, 2])
      qml.CNOT(wires=[2, 0])
      qml.RY(y[0], wires=0)
      qml.RY(y[1], wires=1)
      qml.RY(y[0], wires=2)

  x = pnp.array([[0.2, 0.4, -0.1], [-2.1, 0.5, -0.2], [0.1, 0.7, -0.6]], requires_grad=False)

>   y = pnp.array([1.3, 0.2], requires_grad=True)

Reviewer: Maybe best to keep this as `np.`, since `pnp` is mostly a dev-facing convention?

Author: oops, definitely.

  ```

  ```pycon
  >>> qml.adjoint_metric_tensor(circuit)(x, y)
  tensor([[ 0.25495723, -0.07086695],
          [-0.07086695,  0.24945606]], requires_grad=True)

>   ```

Reviewer: don't forget to add your name below!

<h3>New features since last release</h3>

* Added the adjoint method for the metric tensor.
  [(#1992)](https://github.com/PennyLaneAI/pennylane/pull/1992)

  This method, detailed in [Jones 2020](https://arxiv.org/abs/2011.02991),
  computes the metric tensor using four copies of the state vector and
  a number of operations that scales quadratically in the number of trainable

>   parameters.

Reviewer: out of curioisity, is there a handwavey way of communicating its rough scaling in terms of memory/time?

Author: Included :)

  This method, detailed in [Jones 2020](https://arxiv.org/abs/2011.02991),
  computes the metric tensor using four copies of the state vector and
  a number of operations that scales quadratically in the number of trainable
  parameters (see below for details).
  
  Note that as it makes use of state cloning, it is inherently classical

>   and to be used on state vector simulators and without shots only.

Reviewer: ```suggestion
  and can only be used with statevector simulators and `shots=None`.
```

          [-0.07086695,  0.24945606]], requires_grad=True)
  ```

  Computational cost

  The adjoint method uses :math:`2P^2+4P+1` gates and state cloning operations if the circuit
  is composed only of trainable gates, where :math:`P` is the number of trainable operations.
  If non-trainable gates are included, each of them is applied about :math:`n^2-n` times, where
  :math:`n` is the number of trainable operations that follow after the respective 
  non-trainable operation in the circuit. This means that non-trainable gates later in the 
  circuit are executed less often, making the adjoint method a bit cheaper if such gates

>   appear later.

Reviewer: nice!
